Abort criterion oplus

rainbowsend opened the pull request:
The current implementation uses a fixed number of iterations to update O+ number concentration. This number is controlled by the `nstep_sub`  input option. Instead, I propose to use `nstep_sub` as the maximum number of iterations and break the loop once a certain threshold/epsilon has been reached. This makes the model more versatile, as it automatically adapts the number of iterations to the current situation.

AnonNick committed:
Correct N2+~O collision frequency (#50)

The collision frequency between N2+ and O is misspelled in lamdas. Correct this based on Schunk and Nagy (2009). Thanks to Yizhe Zhang.

rainbowsend committed:
added abort criterion to iteration in oplus, instead of fixed number of iterations

hzfywhn committed:
Reduce code complexity

Equivalent rewrite of the loop to reduce code complexity

hzfywhn commented:
This change looks good to me. It might help saving time during spin-up and other situations when O+ does not change drastically. What do you think? @phamkh @jvilaperez

jvilaperez commented:
It looks good to me too. It can save some iterations in some cases.

hzfywhn committed:
Change converging criteria

As there isn't a thorough test for difference cases, strengthen the criteria for now until we are more confident with the change.

rainbowsend commented:
`calc_terms` is called only at the last iteration
https://github.com/NCAR/tiegcm/blob/a0d5e3323605a47628dd0a64d4ec2e1f3f9d77bb/src/oplus.F#L105-L106
Thus, when the loop is exited before the last iteration, it is never called.

It may be sufficient to move the call outside the loop. Since `calc_terms` only updates `Fe` and `Fn`, and neither variables occur as an argument in any call within the loop, this should be safe. 

Sorry for overlooking this.

hzfywhn committed:
correct exit criteria

The postprocessing terms are calculated at each sub-cycling. Thanks Armin for noticing this.

hzfywhn commented:
> `calc_terms` is called only at the last iteration
> 
> https://github.com/NCAR/tiegcm/blob/a0d5e3323605a47628dd0a64d4ec2e1f3f9d77bb/src/oplus.F#L105-L106
> 
> 
> Thus, when the loop is exited before the last iteration, it is never called.
> It may be sufficient to move the call outside the loop. Since `calc_terms` only updates `Fe` and `Fn`, and neither variables occur as an argument in any call within the loop, this should be safe.
> 
> Sorry for overlooking this.

That is right. We all missed this as well. Just post a quick fix to correct this.

Commit:
compute calc_terms only at last iteration, caclulate error only every 25 iterations to reduce communication costs

rainbowsend commented:
I conducted some tests for the runtime, I will also investigate the differences in O+ later.
For the test, I forwarded a 1 hour IMF run using 60 s integration time and  5° resolution.

The larger the maximal number of iterations, the more runtime is saved when using the abort criterion.

| NSTEP_SUB      | runtime   (fixed, 006bb4e92a531db32b4afe1f8ec101c72e3385a6)  | runtime   (abort, 674d4464eca631d07160f987d9795e31a9f1a8cb) |
| ----------------------- | ------------------------ | ------------------------ |
| 200 |49 s| 46 s
| 400 |79 s| 75 s
| 600 |118 s| 102 s
| 800 |148  s  | 117 s
| 1000 | 178 s | 120 s

Using a less strict abort criterion could further reduce runtime. However, we need to ensure the differences in O+ are reasonable.

hzfywhn commented:
> I conducted some tests for the runtime, I will also investigate the differences in O+ later. For the test, I forwarded a 1 hour IMF run using 60 s integration time and 5° resolution.
> 
> The larger the maximal number of iterations, the more runtime is saved when using the abort criterion.
> 
> NSTEP_SUB	runtime (fixed, [006bb4e](https://github.com/NCAR/tiegcm/commit/006bb4e92a531db32b4afe1f8ec101c72e3385a6))	runtime (abort, [674d446](https://github.com/NCAR/tiegcm/commit/674d4464eca631d07160f987d9795e31a9f1a8cb))
> 200	49 s	46 s
> 400	79 s	75 s
> 600	118 s	102 s
> 800	148 s	117 s
> 1000	178 s	120 s
> Using a less strict abort criterion could further reduce runtime. However, we need to ensure the differences in O+ are reasonable.

Thanks a lot! This benchmark is very helpful.

rainbowsend commented:
I checked how the number of iterations influences O+. For this, I ran the same experiment, varying only the number of iterations (10, 50, 100, 200, 600, 1000, 2000) and not using the abort criterion. The following figure shows the difference in O+ relative to the run with 2000 iterations.
<img width="640" height="700" alt="diff_fixed" src="https://github.com/user-attachments/assets/81111f94-aa14-4738-b7e6-53c29f112848" />.

However, when using the aboart criterion, O+ differences increased with the number of iterations.
So I conducted an experiment in which the loop always exits just one iteration before the maximum number of iterations. With this setup, the differences to the run without an abort criterion are also unrealistically high (above 1E+6 cm-3 in the case of 1000 iterations. For context: Even when using only 10 iterations without abort, the differences in the figure above do not exceed 1000.)

Therefore, the abort criterion should not be used unless the cause of this behaviour is identified. I will check the code and search for dependencies of nsub.

rainbowsend commented:
It turned out that `nstep_sub` is not only used to control the number of iterations but is also used in the calculation of O+:

- in the subroutine `prep_oplus`, which is called before the iteration
https://github.com/NCAR/tiegcm/blob/006bb4e92a531db32b4afe1f8ec101c72e3385a6/src/oplus.F#L568-L570

- in the subroutine `smooth_oplus`, which is called at the beginning of each iteration
https://github.com/NCAR/tiegcm/blob/006bb4e92a531db32b4afe1f8ec101c72e3385a6/src/oplus.F#L674-L678
https://github.com/NCAR/tiegcm/blob/006bb4e92a531db32b4afe1f8ec101c72e3385a6/src/oplus.F#L688-L692

- in the subroutine `iterate_oplus`, which is called in each iteration
https://github.com/NCAR/tiegcm/blob/006bb4e92a531db32b4afe1f8ec101c72e3385a6/src/oplus.F#L902-L905

- in the subroutine `calc_terms`, which is called in the last iteration
https://github.com/NCAR/tiegcm/blob/006bb4e92a531db32b4afe1f8ec101c72e3385a6/src/oplus.F#L1162-L1163

This explains why the results are off when the loop is exited before `nstep_sub` is reached. To exit the loop, the calculations must be performed without depending on `nstep_sub`.
